sync/sync-impl/src/main/java/com/duckduckgo/sync/impl/ui/ShowQRCodeViewModel.kt

```diff
@@ -0,0 +1,100 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.sync.impl.ui
+
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.duckduckgo.anvil.annotations.ContributesViewModel
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.di.scopes.ActivityScope
+import com.duckduckgo.sync.impl.Result.Error
+import com.duckduckgo.sync.impl.Result.Success
+import com.duckduckgo.sync.impl.SyncRepository
+import com.duckduckgo.sync.impl.ui.ShowQRCodeViewModel.Command.LoginSucess
+import com.duckduckgo.sync.impl.ui.SyncInitialSetupViewModel.Command
+import javax.inject.*
+import kotlinx.coroutines.channels.BufferOverflow.DROP_OLDEST
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.receiveAsFlow
+import kotlinx.coroutines.launch
+
+@ContributesViewModel(ActivityScope::class)
+class ShowQRCodeViewModel @Inject constructor(
+    private val syncRepository: SyncRepository,
+    private val dispatchers: DispatcherProvider,
+) : ViewModel() {
+    private val command = Channel<Command>(1, DROP_OLDEST)
+    private val viewState = MutableStateFlow(ViewState())
+
+    fun viewState(): Flow<ViewState> = viewState.onStart {
+        startPolling()
+    }
+
+    private fun startPolling() {
+        viewModelScope.launch(dispatchers.io()) {
+            showQRCode()
+            var polling = true
+            while (polling) {
+                delay(POLLING_INTERVAL)
+                when (syncRepository.pollConnectionKeys()) {
+                    is Success -> {
+                        command.send(LoginSucess)
+                        polling = false
+                    }
+                    else -> {
+                        // noop - keep polling
+                    }
+                }
+            }
+        }
+    }
+
+    fun commands(): Flow<Command> = command.receiveAsFlow()
+
+    data class ViewState(
+        val qrCode: String? = null,
+    )
+
+    sealed class Command {
+        object LoginSucess : Command()
+        object Error : Command()
+    }
+
+    private suspend fun showQRCode() {
+        val result = syncRepository.getConnectQR()
+        when (result) {
+            is Error -> {
+                command.send(Command.Error)
+            }
+            is Success -> {
+                viewState.emit(
+                    viewState.value.copy(
+                        qrCode = result.data,
+                    ),
+                )
+            }
+        }
+    }
+
+    companion object {
+        const val POLLING_INTERVAL = 5000L
```

Comment:
Something that works. I've shared this with others just in case I'm missing something or maybe others prefer to agree on using the same value across clients. imho, this is an implementation detail, but task here: https://app.asana.com/0/1201493110486074/1204280397189535/f